bpo-41631: Check that importing _ast returns the right module

encukou opened the pull request:
Hopefully #21961 will be OK, but if not: here is a relatively minimal fix for [bpo-41631](https://bugs.python.org/issue41631), which basically makes compiling to/from AST fail gracefully (with a proper exception rather than a NULL without exception) if the `_ast` module has been tampered with.
For example, it looks like Mercurial's lazy-loading scheme replaces `_ast`with a proxy object; this will make AST unusable.

The second commit deals with a hack in pegen's tests – please review the commits separately, and see comments.


<!-- issue-number: [bpo-41631](https://bugs.python.org/issue41631) -->
https://bugs.python.org/issue41631
<!-- /issue-number -->

encukou committed:
bpo-41631: Check that importing _ast returns the right module

encukou committed:
Work around a hack in pegen's tests, an extension that must be treated as _ast

pablogsal commented on Tools/peg_generator/pegen/build.py:
```diff
@@ -59,6 +59,12 @@ def compile_c_extension(
+
+    # testing pegen requires generating a C extension module, which contains
+    # a copy of the symbols defined in Python-ast.c and needs to be treated
+    # as the _ast module. See ast_dump.py.
```
```suggestion
    # testing pegen requires generating C extension modules, which contains
    # a copy of the symbols defined in Python-ast.c and needs to be treated
    # as the _ast module. See ast_dump.py.
```

encukou committed:
Update Tools/peg_generator/pegen/build.py

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

encukou committed:
Don't mention "segfault" from the blurb

The previous behavior didn't always segfault -- in fact, it usually
raised a SystemError as well. A segfault (or any other behavior,
depending on the module object imported as _ast) was still *possible*,
but for that level of detail people should see the bpo issue.
